crates/red_knot/src/types/infer.rs

```diff
@@ -16,21 +16,30 @@ use crate::{FileId, Name};
 
 // FIXME: Figure out proper dead-lock free synchronisation now that this takes `&db` instead of `&mut db`.
 #[tracing::instrument(level = "trace", skip(db))]
-pub fn infer_symbol_type(db: &dyn SemanticDb, symbol: GlobalSymbolId) -> QueryResult<Type> {
+pub fn infer_symbol_public_type(db: &dyn SemanticDb, symbol: GlobalSymbolId) -> QueryResult<Type> {
     let symbols = symbol_table(db, symbol.file_id)?;
     let defs = symbols.definitions(symbol.symbol_id);
     let jar: &SemanticJar = db.jar()?;
 
-    if let Some(ty) = jar.type_store.get_cached_symbol_type(symbol) {
+    if let Some(ty) = jar.type_store.get_cached_symbol_public_type(symbol) {
         return Ok(ty);
     }
 
-    // TODO handle multiple defs, conditional defs...
-    assert_eq!(defs.len(), 1);
-
-    let ty = infer_definition_type(db, symbol, defs[0].clone())?;
+    // The public type of a symbol is the union of all of its definitions. This is the most
+    // cautious/sound approach, though it can lead to a broader-than-desired type in a case like
+    // `x = 1; x = str(x)`, where the first definition of `x` can never be visible. TODO prune
+    // definitions that we can prove can't be visible.
+    let tys = defs
+        .iter()
+        .map(|def| infer_definition_type(db, symbol, def.clone()))
+        .collect::<QueryResult<Vec<Type>>>()?;
+    let ty = match tys.len() {
+        0 => Type::Unknown,
+        1 => tys[0],
+        _ => Type::Union(jar.type_store.add_union(symbol.file_id, &tys)),
+    };
```

Comment:
```suggestion
    let ty = match &*tys {
        [] => Type::Unknown,
        [ty] => ty,
        tys => Type::Union(jar.type_store.add_union(symbol.file_id, &tys)),
    };
```

Reply:
Not using this because I'm using the version that avoids collecting in the common case, but this is also a really good lesson in the power of `match`, thank you :)